src/peer_loop.rs

```diff
@@ -1300,17 +1301,44 @@ impl PeerLoopHandler {
                     return Ok(KEEP_CONNECTION_ALIVE);
                 }
 
+                // 6. if transaction double spends, punish.
+                let double_spends = transaction
+                    .kernel
+                    .double_spends(&mutator_set_accumulator_after);
+                if double_spends {
+                    warn!("Received double-spending tx");
+                    self.punish(NegativePeerSanction::DoubleSpendingTransaction)
+                        .await?;
+                    return Ok(KEEP_CONNECTION_ALIVE);
+                }
+
+                // 7. If transaction cannot be applied to mutator set, punish. I'm not sure how this
+                // could happen when above checks pass.
+                let ms_update = MutatorSetUpdate::new(
+                    transaction.kernel.inputs.clone(),
+                    transaction.kernel.outputs.clone(),
+                );
+                let can_apply = ms_update
+                    .apply_to_accumulator(&mut mutator_set_accumulator_after.clone())
+                    .is_ok();
```

Comment:
I don't see what's wrong with these six lines but I just want to point out the (afaik) new and unorthodox use of `MutatorSetUpdate`. Previously it was only ever used in relation to blocks, not individual standalone transactions. (Note that `mutator_set_update.rs` lives in `block/` not `mutator_set/`.)